substrate/client/rpc-spec-v2/src/archive/archive.rs

```diff
@@ -278,4 +302,92 @@ where
 
 		Ok(storage_client.handle_query(hash, items, child_trie))
 	}
+
+	fn archive_unstable_storage_diff(
+		&self,
+		pending: PendingSubscriptionSink,
+		hash: Block::Hash,
+		items: Vec<ArchiveStorageDiffItem<String>>,
+		previous_hash: Option<Block::Hash>,
+	) {
+		let storage_client = ArchiveStorageDiff::new(self.client.clone());
+		let client = self.client.clone();
+
+		log::trace!(target: LOG_TARGET, "Storage diff subscription started");
+
+		let fut = async move {
+			let Ok(mut sink) = pending.accept().await.map(Subscription::from) else { return };
+
+			// Deduplicate the items.
+			let mut trie_items = match deduplicate_storage_diff_items(items) {
+				Ok(items) => items,
+				Err(error) => {
+					let _ = sink.send(&ArchiveStorageDiffEvent::err(error.to_string())).await;
+					return
+				},
+			};
+			// Default to using the main storage trie if no items are provided.
+			if trie_items.is_empty() {
+				trie_items.push(Vec::new());
+			}
+			log::trace!(target: LOG_TARGET, "Storage diff deduplicated items: {:?}", trie_items);
+
+			let previous_hash = if let Some(previous_hash) = previous_hash {
+				previous_hash
+			} else {
+				let Ok(Some(current_header)) = client.header(hash) else {
+					let message = format!("Block header is not present: {hash}");
+					let _ = sink.send(&ArchiveStorageDiffEvent::err(message)).await;
+					return
+				};
+				*current_header.parent_hash()
+			};
+
+			let (tx, mut rx) = tokio::sync::mpsc::channel(STORAGE_QUERY_BUF);
+			for trie_queries in trie_items {
+				let storage_fut = storage_client.handle_trie_queries(
+					hash,
+					previous_hash,
+					trie_queries,
+					tx.clone(),
+				);
+				let result =
+					futures::future::join(storage_fut, process_events(&mut rx, &mut sink)).await;
+				if !result.1 {
+					log::debug!(target: LOG_TARGET, "Error processing trie queries");
+					return;
+				}
+			}
+
+			let _ = sink.send(&ArchiveStorageDiffEvent::StorageDiffDone).await;
+		};
+
+		self.executor.spawn("substrate-rpc-subscription", Some("rpc"), fut.boxed());
+	}
+}
+
+/// Returns true if the events where processed successfully, false otherwise.
+async fn process_events(
+	rx: &mut mpsc::Receiver<ArchiveStorageDiffEvent>,
+	sink: &mut Subscription,
+) -> bool {
```

Comment:
See my comment above, would be neat to remove the bool flag, process all trie queries in `handle_trie_queries` and just deal with logs, sending out all message and so in this function.